src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -300,22 +301,32 @@ public int Available
 
                 if (_localEndPoint == null)
                 {
-                    Internals.SocketAddress socketAddress = IPEndPointExtensions.Serialize(_rightEndPoint);
+                    Span<byte> buffer = stackalloc byte[SocketAddress.GetMaximumAddressSize(_addressFamily)];
+                    int size = buffer.Length;
+                    // This may throw ObjectDisposedException.
 
                     unsafe
                     {
-                        fixed (byte* buffer = socketAddress.InternalBuffer)
-                        fixed (int* bufferSize = &socketAddress.InternalSize)
+                        fixed (byte* ptr = &MemoryMarshal.GetReference(buffer))
                         {
-                            // This may throw ObjectDisposedException.
-                            SocketError errorCode = SocketPal.GetSockName(_handle, buffer, bufferSize);
+                            SocketError errorCode = SocketPal.GetSockName(_handle, ptr, &size);
                             if (errorCode != SocketError.Success)
                             {
                                 UpdateStatusAfterSocketErrorAndThrowException(errorCode);
                             }
                         }
                     }
-                    _localEndPoint = _rightEndPoint.Create(socketAddress);
+
+                    if (_addressFamily == AddressFamily.InterNetwork || _addressFamily == AddressFamily.InterNetworkV6)
+                    {
+                        _localEndPoint = IPEndPointExtensions.CreateIPEndPoint(buffer.Slice(0, size));
+                    }
+                    else
+                    {
+                        SocketAddress socketAddress = new SocketAddress(_rightEndPoint.AddressFamily, size);
+                        buffer.Slice(0, size).CopyTo(socketAddress.Buffer.Span);
+                        _localEndPoint = _rightEndPoint.Create(socketAddress);
```

Comment:
We should really consider renaming "_rightEndPoint".  Doesn't need to be as part of this PR, but it's not at all clear what "right" means.

Reply:
yah. I was thinking about if for a long time. And we really only need it IMHO if we deal with strange socket types .... which is rare IMHO.

Reply:
btw if you have good name, please let me know.

Reply:
I don't even know what it is, even after reading the comment on it :) Can you describe its purpose?

Reply:
I'll put it ion 9.0 TODO cleanup list. There still may be some trickery around DualMode sockets but I don't know fir sure. AFAIK it allows to create EndPoints for AddressTypes .NET does not support internally. We can perhaps also look at .NET Framework to see if there is some legacy for it.